scripts/azdo: add AZDO component conf files

amstewart opened the pull request:
Add an auto.conf file, containing bitbake configuration settings which
(a) are common to all NI-internal AZDO bitbake components and (b) have
static values.

Add an ni-org.conf file, containing bitbake configuration settings which
are necessary to produce faster, more-reproducible builds, but which are
insane for an external (non-NI) builder.

Signed-off-by: Alex Stewart <alex.stewart@ni.com>

----

### Testing

* Rebuilt `rtos_oe_feeds` on my dev machine and verified that the variables I moved have their correct values and that all makefile targets complete.

### Associated PRs
* meta-nilrt [PR 125](https://github.com/ni/meta-nilrt/pull/125) for the nilrt.git angle on this change.
* [AZDO](https://dev.azure.com/ni/DevCentral/_git/ni-central/pullrequest/123493)

amstewart committed:
scripts/azdo: add AZDO component conf files

Add an auto.conf file, containing bitbake configuration settings which
(a) are common to all NI-internal AZDO bitbake components and (b) have
static values.

Add an ni-org.conf file, containing bitbake configuration settings which
are necessary to produce faster, more-reproducible builds, but which are
insane for an external (non-NI) builder.

Signed-off-by: Alex Stewart <alex.stewart@ni.com>

amstewart commented:
@ni/rtos

amstewart committed:
auto.conf: enable cve-check

The team agrees that now is the time to enable cve-check on all of our
autobuilds. Enable it.

Signed-off-by: Alex Stewart <alex.stewart@ni.com>

shruthi-ravi commented on scripts/azdo/conf/auto.conf:
```diff
@@ -0,0 +1,14 @@
+include conf/ni-org.conf
```
I'm not convinced that this needs to be a separate conf file from ni-org.conf. This too will not be used by anyone outside NI, right? 
I'm also concerned that for a developer build outside azdo, which is a common-enough case, the configuration in these two files will not be included in the build by default. What are your thoughts about that?

amstewart replied:
I broke out configuration into the auto.conf which enabled unnecessary build artifacts (buildhistory, cve logs, et c.). My thinking was that NI devs can use the auto.conf, but not waste time making those extraneous artifacts, unless they want to explicitly add them to their local.conf.

> This too will not be used by anyone outside NI, right?

Correct.

>I'm also concerned that for a developer build outside azdo, which is a common-enough case, the configuration in these two files will not be included in the build by default. What are your thoughts about that?

I suppose we can add a switch to [`nilrt-build-init.env`](https://github.com/ni/nilrt/blob/nilrt/master/dunfell/nilrt-build-init.env) which adds the ni-org.conf to your build environment automatically. Would that suffice?

shruthi-ravi replied:
> I suppose we can add a switch to nilrt-build-init.env which adds the ni-org.conf to your build environment automatically. Would that suffice?

I guess there aren't better alternatives. Can you add a note to the README recommending how to invoke the script for NI developers? And announce this change to the team.

amstewart committed:
nilrt-init-build.env: add switch to enable org.conf

Add a switch to the nilrt-init-build helper script which will add the
ni-org.conf configuration snippet to the build directory as site.conf.

The switch should be asserted for all builders which are internal to the
NI corporate network.

Signed-off-by: Alex Stewart <alex.stewart@ni.com>

amstewart commented:
PATCH V2:
* added a switch (`-o` | `--org`) to `nilrt-build-init.env`, which will add the `ni-org.conf` snippet as `conf/site.conf`, for use by developers. External builders can still just source the init file with no arguments, and it will work in the same way as before.

billpittman commented on nilrt-build-init.env:
```diff
@@ -2,6 +2,21 @@
 
 SCRIPT_ROOT=$(realpath $(dirname ${BASH_SOURCE:-${0}}))
 
+enable_ni_org_conf=false
+positionals=()
+
+while [ $# -ge 1 ]; do case "$1" in
+	-o|--org)
+		enable_ni_org_conf=true
```
I'm torn, the selfish side of me wants our use case to be the "no parameters - just add in the NI stuff".  However I also recognize that users will "always" fail if they just invoke the script without parameters.  So I understand why you did the logic the way you did.

amstewart replied:
I figure NI developers are smart enough to know how to alias or wrap the script in their workspace to automatically use `--org`.

shruthi-ravi commented on nilrt-build-init.env:
```diff
@@ -2,6 +2,21 @@
+
+while [ $# -ge 1 ]; do case "$1" in
+	-o|--org)
+		enable_ni_org_conf=true
```
I found the commit wording confusing and my bash knowledge is minimal, so let me confirm my understanding here. An NI user now has to invoke this script with the -o parameter `. nilrt-build-init.env -o` to copy the ni-org.conf file to the build directory. Correct?

billpittman replied:
shruthi - yes that's my read, and I'm not completely happy with that solution.

@amstewart Could we possibly make the script a bit smarter and check the user's email address and if its an @ni.com email, initialize enable_ni_org_conf=true ?  'git config user.email |grep "\@ni.com$"'  - or some such ??  That way the script would basically do the right thing most of the time (assuming we all have (or set) our email addresses with git).

amstewart replied:
I don't like the idea of the script trying to be smart like that. Having an NI email address configured doesn't _mean_ that you are connected to the corporate network. I disconnect from the VPN all the time and still want to do bitbake things, while my repo still has an NI email attached to it.

I've considered a few different heuristics for determining if you're connected to the corporate network. But none of them really get around the fundamental requirement that VPN builders understand that these config settings are a thing, and that they will only work while connected to the VPN.

I think its an improvement that they are broken out specifically from the rest of the local.conf. And I like the implication of the `-o` switch on the initscript; it implies "Yes, I understand the implications of adding these config settings and want to be responsible for turning them off again, if I need to."

As I said above, you can just wrap the script on your local machine with a one-line source file that provides the `-o` argument, or edit the initscript locally.

billpittman replied:
Too true, you've got a couple good points there.  I guess I'll have to write a script or remember something ;)

shruthi-ravi reviewed (approved):
lgtm as long as changes are communicated to the team so that noone is caught by surprise.

amstewart committed:
README: update README with new initscript argument

Signed-off-by: Alex Stewart <alex.stewart@ni.com>

amstewart commented:
### Patch V3
* Updated the `README.md` to advertise the new `--org` argument and added a section telling NI developers to use it.

amstewart commented:
### Patch V4
* added a fixup to the `auto.conf` script to remove the `include conf/ni-org.conf` line. I decided to install the `ni-org.conf` script as `conf/site.conf` in the AZDO components, which is automatically included, so the includes line is unnecessary.
